lnwallet/channel.go

```diff
@@ -2985,21 +2927,50 @@ func (lc *LightningChannel) evaluateHTLCView(view *HtlcView, ourBalance,
 				entry, ourBalance, theirBalance,
 				whoseCommitChain, true,
 			)
-
-			// If we are mutating the state, then set the add
-			// height for the appropriate commitment chain to the
-			// next height.
-			if mutateState {
-				addHeights.SetForParty(
-					whoseCommitChain, nextHeight,
-				)
-			}
 		}
 
 		newView.TheirUpdates = append(newView.TheirUpdates, entry)
 	}
 
-	return newView, nil
+	// Create a function that is capable of identifying whether or not the
+	// paymentDescriptor has been committed in the commitment chain
+	// corresponding to whoseCommitmentChain.
+	isUncommitted := func(update *paymentDescriptor) bool {
+		switch update.EntryType {
+		case Add:
+			h := update.addCommitHeights.GetForParty(
+				whoseCommitChain,
+			)
+
+			return h == 0
+
+		case FeeUpdate:
+			h := update.addCommitHeights.GetForParty(
+				whoseCommitChain,
+			)
+
+			return h == 0
+
+		case Settle, Fail, MalformedFail:
+			h := update.removeCommitHeights.GetForParty(
+				whoseCommitChain,
+			)
+
+			return h == 0
+
+		default:
+			panic("invalid paymentDescriptor EntryType")
+		}
+	}
+
+	// Collect all of the updates that haven't had their commit heights sent
+	// for the commitment chain corresponding to whoseCommitmentChain.
+	uncommittedUpdates := lntypes.Dual[[]*paymentDescriptor]{
+		Local:  fn.Filter(isUncommitted, view.OurUpdates),
+		Remote: fn.Filter(isUncommitted, view.TheirUpdates),
+	}
+
+	return newView, uncommittedUpdates, nil
```

Comment:
Seems like an improvement. I'll tack on a commit to the end.

Reply:
So I went back and tried to do this and here's the issue. Notice how the thing we are filtering in the cited code is the uncommitted updates on `view`, *not* on `newView`. The process that occurs above this code specifically pairs off uncommitted settles and removes the corresponding add. In essence this evaluate view is a compression function and the compression loses the info we would need to subsequently compute the uncommitted settles. We *could* augment the view here to not lose that information so we can query it later, but I'm not thoroughly convinced that that's the right move. I'm open to input here.

The reason we need to keep the uncommitted settle information is so that we can properly record the lifetime sent/received accumulators in the view. If we don't have the uncommitted settles, we can't traverse them and accumulate their effects directly. Now we can *probably* repurpose the balance deltas to give us that information on. This is technically not semantics invariant the way the rest of the refactor series is, but it may be fine to do anyway. Lmk what you think here.

Reply:
I can definitely put it as a method on the view, I just can't eliminate the need for the extra return value here since we need to report the pre-fold updates.